src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -549,18 +644,32 @@ void ReplicatedMergeTreeSink::commitPart(
             throw Exception(ErrorCodes::QUERY_WAS_CANCELLED,
                             "Insert query (for block {}) was cancelled by concurrent ALTER PARTITION", block_number_lock->getPath());
         }
-        else if (multi_code == Coordination::Error::ZCONNECTIONLOSS
-            || multi_code == Coordination::Error::ZOPERATIONTIMEOUT)
+        else if (Coordination::isHardwareError(multi_code))
```

Comment:
I wonder if it would be possible to improve this error and the use of `part_committed_locally_but_zookeeper`.

Currently if there is a hardware error here CH checks if the part was written to ZK, if it is then it returns success and if it doesn't it returns a failure `UNEXPECTED_ZOOKEEPER_ERROR`, asking the customer to retry and leaving the part to be cleaned up later.

Knowing that of the operations done in the process is to unlock the `block_number_lock`, could we use the that to retry these operations if the lock still exist? The idea would be to retry only this request and not the whole function. WDYT @devcrafter ?

One thing that might be necessary to do this would be to break down the function a big (it's over 600 lines with lots of branches) so it's simpler to decide if ZK retrial, and which part, is possible. It might require a refactor of the function and of the retries (to share retries across different blocks).

Reply:
> Knowing that of the operations done in the process is to unlock the `block_number_lock`, could we use the that to retry these operations if the lock still exist? The idea would be to retry only this request and not the whole function. 

`block_number_lock` creates an ephemeral node in zookeeper. After the connection/session to the zookeeper is lost/closed, it's no longer valid, and so `block_number`. So, this way is not viable (until I miss something)

Reply:
Usually the ephemeral lock will still be alive as ZK keeps it for a set amount of time so we can try to recover it. IIRC for the table locks, we check if it exists and if it does delete it and recreate it to gain it again thus avoiding the need to wait until deletion happens normally. We could do something similar as long as we keep it concurrent safe.